pennylane/measurements/mid_measure.py

```diff
@@ -209,22 +210,55 @@ def func(x):
               samples, leading to unexpected or incorrect results.
 
     """
-    wire = Wires(wires)
-    if len(wire) > 1:
+    if qml.capture.enabled():
+        primitive = _create_mid_measure_primitive()
+        return primitive.bind(wires, reset=reset, postselect=postselect)
+
+    return _measure_impl(wires, reset=reset, postselect=postselect)
+
+
+def _measure_impl(
+    wires: Union[Hashable, Wires], reset: Optional[bool] = False, postselect: Optional[int] = None
+):
+    """Concrete implementation of qml.measure"""
+    wires = Wires(wires)
+    if len(wires) > 1:
         raise qml.QuantumFunctionError(
             "Only a single qubit can be measured in the middle of the circuit"
         )
 
     # Create a UUID and a map between MP and MV to support serialization
     measurement_id = str(uuid.uuid4())[:8]
-    mp = MidMeasureMP(wires=wire, reset=reset, postselect=postselect, id=measurement_id)
-    if qml.capture.enabled():
-        raise NotImplementedError(
-            "Capture cannot currently handle classical output from mid circuit measurements."
-        )
+    mp = MidMeasureMP(wires=wires, reset=reset, postselect=postselect, id=measurement_id)
     return MeasurementValue([mp], processing_fn=lambda v: v)
 
 
+@lru_cache
+def _create_mid_measure_primitive():
+    """Create a primitive corresponding to an mid-circuit measurement type.
+
+    Called when using :func:`~pennylane.measure`.
+
+    Returns:
+        jax.core.Primitive: A new jax primitive corresponding to a mid-circuit
+        measurement.
+
+    """
+    import jax  # pylint: disable=import-outside-toplevel
+
+    mid_measure_p = jax.core.Primitive("measure")
+
+    @mid_measure_p.def_impl
+    def _(wires, reset=False, postselect=None):
+        return _measure_impl(wires, reset=reset, postselect=postselect)
+
+    @mid_measure_p.def_abstract_eval
+    def _(*_, **__):
+        return jax.core.ShapedArray((), jax.numpy.bool_)
```

Comment:
Open question: Catalyst uses boolean tracers for abstract evals of mid-circuit measurements. However, in PennyLane, the current behaviour of measurement values is more consistent with integers. Eg, `m0 + m1` will behave like integer addition rather than a logical OR operation. I'm not sure how to reconcile this behaviour. cc @trbromley @josh146

Reply:
@mudit2812 I can't explain why, but `m0 + m1` works fine in Catalyst:

```python
dev = qml.device("lightning.qubit", wires=3)

@qml.qjit(autograph=True)
@qml.qnode(dev)
def circuit(x):

    def prepare_and_measure(x):
        m = jnp.zeros([2])

        for i in range(2):
            qml.RX(x, wires=i)
            m[i] = catalyst.measure(i)

        return m

    m0, m1 = prepare_and_measure(x)

    while m0 + m1 != 2:
        m0, m1 = prepare_and_measure(x)

    qml.RY(m0 + m1, wires=2)
    return {"m0": m0, "m1": m1, "probs2": qml.probs(wires=2)}
```

```pycon
>>> circuit(0.9)
{'m0': Array(1., dtype=float64),
 'm1': Array(1., dtype=float64),
 'probs2': Array([0.29192658, 0.70807342], dtype=float64)}
```

Could be because, in Python, `True + True == 2`? E.g., the casting is implicit. But may tag @dime10 in case he knows.

Reply:
So Catalyst _already_ behaves like PL regarding integer addition of measurements

Reply:
Would be curious if there is any edge case where the behaviours differ :thinking: It looks like Python generally promotes Boolean arithmetic to integers.

Reply:
I hadn't actually executed it (since I can't), but in the jaxpr, with a boolean tracer, a logical OR and addition have the same repr for the primitive, hence my question:

```python
import pennylane as qml
import numpy as np
import jax
import jax.numpy as jnp

qml.capture.enable()

@qml.qnode(qml.device("default.qubit", wires=2, shots=10))
def f(x):
    m1 = qml.measure(0)
    m2 = qml.measure(0, reset=True)
    m3 = m1 + m2
    m4 = m1 | m2
    return qml.sample([m1, m2]), qml.expval(m4)

print(jaxpr := jax.make_jaxpr(f.func)(1.0))
```
```pycon
{ lambda ; a:f64[]. let
    b:f64[] = mul 2.0 a
    _:AbstractOperator() = RX[n_wires=1] b 0
    c:bool[] = measure[postselect=None reset=False] 0
    d:bool[] = measure[postselect=None reset=True] 0
    _:bool[] = or c d
    e:bool[] = or c d
    f:AbstractMeasurement(n_wires=2) = sample_mcm[single_mcm=False] c d
    g:AbstractMeasurement(n_wires=1) = expval_mcm[single_mcm=True] e
  in (f, g) }
```

Reply:
Erick figured out the reason Josh's example works is because it stores the MCM in the array `jnp.zeros(2)`, which apparently defaults to a float type 😅

Reply:
I found that while Python will treat the addition of Booleans like an integer addition, NumPy does not. Instead, it interprets a `+` like a logical OR, `*` like a logical AND, does not allow `-` (hinting `^` should be used instead), `^` as a logical XOR, and `/` actually produces a float :O 

Why I'm not sure, but JAX will probably follow NumPy rules on this. So I guess our decision depends on whether we want to present users with Boolean logic or integer arithmetic semantics on MCM results by default (they can always cast from one to the other). 
Personally I'm okay with either, and we can do what's most convenient for the "general" use case. @josh146 @trbromley Any thoughts here?

Reply:
I would suggest we use integer types by default (for now). This is partly because without _any_ processing, I would expect a measurement `m` to be of type `0` or `1` by default.